File: 2025/day1/main.py
from pathlib import Path

import pytest

TEST_INPUT = """\
L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
"""

EXPECTED_PART1 = 3

EXPECTED_PART2 = 6


def read_input_file(file_path: Path = Path("input")):
    with open(file_path, "r") as fp:
        data = fp.read()
    return data


def part1(contents: str, start_pos: int = 50) -> int:
    dial = start_pos
    count = 0
    for line in contents.splitlines():
        direction = line[0]
        num = int(line[1:])
        if direction == "L":
            num *= -1
        dial = (dial + num) % 100
        if dial == 0:
            count += 1
    return count


def part2(contents: str, start_pos: int = 50) -> int:
    # Brute force
    dial = start_pos
    count = 0
    for line in contents.splitlines():
        direction = line[0]
        num = int(line[1:])
        for i in range(num):
            if direction == "L":
                dial -= 1
            else:
                dial += 1
            dial %= 100
            if dial == 0:
                count += 1
    return count


@pytest.mark.parametrize(
    ("test_input", "expected"),
    ((TEST_INPUT, EXPECTED_PART1),),
)
def test_part1(test_input: str, expected: int) -> None:
    assert part1(test_input) == expected


@pytest.mark.parametrize(
    ("test_input", "expected"),
    ((TEST_INPUT, EXPECTED_PART2),),
)
def test_part2(test_input: str, expected: int) -> None:
    assert part2(test_input) == expected


if __name__ == "__main__":
    contents = read_input_file("input")

    answer = part1(contents)
    print(answer)

    answer = part2(contents)
    print(answer)

File: 2025/day1/src/main.rs
use std::{fs, ops::Rem};

const TEST_INPUT: &str = "\
L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
";
const EXPECTED_PART1: i32 = 3;
const EXPECTED_PART2: i32 = 6;

fn read_input_file(file_path: &str) -> String {
    let contents = fs::read_to_string(file_path).unwrap();
    contents
}

fn part1(contents: &str) -> i32 {
    let mut dial = 50;
    let mut count = 0;

    for line in contents.lines() {
        let dir = line.chars().next().unwrap();
        let click = line[1..].parse::<i32>().unwrap();
        let delta = if dir == 'L' { -click } else { click };
        dial = (dial + delta).rem_euclid(100);
        if dial == 0 {
            count += 1;
        }
    }
    count
}

fn part2(contents: &str) -> i32 {
    let mut dial: i32 = 50;
    let mut count: i32 = 0;

    for line in contents.lines() {
        let dir = line.chars().next().unwrap();
        let click = line[1..].parse::<i32>().unwrap();
        let step = if dir == 'L' { -1 } else { 1 };
        for _ in 0..click {
            dial = (dial + step).rem_euclid(100);
            if dial == 0 {
                count += 1;
            }
        }
    }
    count
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_part1() {
        let results: i32 = part1(TEST_INPUT);
        assert_eq!(results, EXPECTED_PART1)
    }

    #[test]
    fn test_part2() {
        let results: i32 = part2(TEST_INPUT);
        assert_eq!(results, EXPECTED_PART2)
    }
}

fn main() {
    let contents = read_input_file("input");

    let answer = part1(&contents);
    println!("{answer}");

    let answer = part2(&contents);
    println!("{answer}");
}
